pkg/operator/metricscontroller/legacy_cn_certs_sync.go

```diff
@@ -0,0 +1,58 @@
+package metricscontroller
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	prometheusv1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+)
+
+// NewLegacyCNCertsMetricsSyncFunc creates a metrics sync function that executes the given query
+// and interprets the result value as the count of invalid certificates containing CN fields used as host names.
+// If the value is >=1 the `InvalidCertsDetected` condition state will be set.
+//
+// The supported prometheus query result types are:
+// - vector: only the first sample of the vector will be used for evaluation.
+// - scalar: the returned value will be used.
+func NewLegacyCNCertsMetricsSyncFunc(query string, operatorClient v1helpers.OperatorClient) MetricsSyncFunc {
```

Comment:
this doesn't look very useful to expose.

I think people wanted a, "pass a cert or a secret along with a servername and determine if it was valid or not"

Reply:
This is supposed to leverage the metric exposed via https://github.com/kubernetes/kubernetes/pull/95396.

Reply:
And a metric we are introducing in oauth-server: https://github.com/openshift/enhancements/blob/9225dbdfa07d7479f412e6d2c5f275f484def5b0/enhancements/authentication/invalid-certs.md#oauth-server

Reply:
Having said that, the idea to fetch secrets directly in the operator seems viable as well. Let me check if we can access all of this in the respective operators.

Reply:
@stlaz as far as i know, the actual target server certificates are not available as secrets and/or configmaps, are they?

Reply:
alternatively we could, redundantly implement the metrics sync function in the corresponding operators as well, as they have intimate knowledge over the underlying query and their return type, hence the implementation would vary slightly as well.

(in case they don't expose the server cert via config/secret)

Reply:
>  as far as i know, the actual target server certificates are not available as secrets and/or configmaps, are they?

In case of the metrics, they aren't. You would only know which CA bundle to use to connect to the webhook/aggregated API, but that does not help you much because CA bundles are irrelevant to the SAN checks.

What I think David is trying to point out is that it would be useful to have a validation function you can use in places where the server certificate would in fact appear in secrets/configmaps (=> config).

Reply:
> validation function you can use in places where the server certificate would in fact appear in secrets/configmaps (=> config).

agreed, however i want to make this as a separate PR.

Reply:
I'm not against using a metric  in your controller (though I probably would not have chose that).  I'm saying that the request for an exported library function was about inspecting a cert, not making a prom query that assumes other code has inspected.

Reply:
As clarified out of band: this is not supposed to replace the idea of introducing functions or a controller to validate certificates explicitly. This functionality will be provided as a separate PR.